ee/tabby-webserver/src/oauth/general.rs

```diff
@@ -0,0 +1,205 @@
+}
+
+#[async_trait]
+impl OAuthClient for GeneralClient {
+    async fn exchange_code_for_token(&self, code: String) -> Result<String> {
+        let credential = self.read_credential().await?;
+        let redirect_uri = self.auth.oauth_callback_url(OAuthProvider::General).await?;
+        let params: [(&str, &str); 5] = [
+            ("client_id", &credential.client_id),
+            ("client_secret", &credential.client_secret),
+            ("code", &code),
+            ("grant_type", "authorization_code"),
+            ("redirect_uri", &redirect_uri),
+        ];
+
+        let config_url = credential.config_url;
+        let oidc_config = self.retrieve_oidc_config(config_url).await;
+        let token_endpoint = oidc_config.token_endpoint;
+
+        let token = self
+            .client
+            .post(token_endpoint)
+            .header(reqwest::header::ACCEPT, "application/json")
+            .form(&params)
+            .send()
+            .await?
+            .json::<GeneralOAuthResponse>()
+            .await?;
+
+        if token.access_token.is_empty() {
+            bail!(
+                "Failed to exchange access token: {}",
+                token.error_description
+            );
+        }
+
+        Ok(token.access_token)
+    }
+
+    async fn fetch_user_email(&self, access_token: &str) -> Result<String> {
+        let user_info = self.retrieve_user_info(access_token).await?;
+        if user_info.email.is_empty() {
+            bail!("No email found in user info");
+        }
+
+        Ok(user_info.email)
+    }
+
+    async fn fetch_user_full_name(&self, access_token: &str) -> Result<String> {
+        let user_info = self.retrieve_user_info(access_token).await?;
+        if user_info.name.is_empty() {
+            bail!("No name found in user info");
+        }
+
+        Ok(user_info.name)
+    }
+
+    async fn get_authorization_url(&self) -> Result<String> {
+        let credential = self.read_credential().await?;
+
+        let config_url = credential.config_url;
+        let oidc_config = self.retrieve_oidc_config(config_url).await;
+        let authorization_endpoint = &oidc_config.authorization_endpoint;
+
+        let scope = oidc_config.scopes_supported.join(" ");
+        let redirect_uri = &self.auth.oauth_callback_url(OAuthProvider::General).await?;
+
+        let mut url = reqwest::Url::parse(authorization_endpoint)?;
+
+        let params: [(&str, &str); 4] = [
+            ("client_id", &credential.client_id),
+            ("response_type", "code"),
+            ("scope", &scope),
+            ("redirect_uri", redirect_uri),
+        ];
+        for (k, v) in params {
+            url.query_pairs_mut().append_pair(k, v);
+        }
+
+       Ok(url.to_string())
+    }
+}
+
+#[cached(
+    type = "TimedCache<String, OAuthConfig>",
+    create = "{ TimedCache::with_lifespan(3600 * 12) }",
+    convert = r#"{ url.to_string() }"#
+)]
+async fn retrieve_oidc_config_cached(url: String) -> OAuthConfig {
```

Comment:
Note: I created this method to cache the outgoing calls to `https://server.com/.well-known/openid-configuration`. This endpoint contains a JSON payload containing the authorization, user info, and token endpoints and well as the permission scopes.